[19기 김남선] Modify : Bids 코드 리뷰 수정사항 반영

kimnamsun opened the pull request:
>wecode PR message template 및 체크 리스트 입니다. 
아래 사항들을 전부 작성/체크 하시고 PR 해주세요!

## 수정 사항 간략한 한줄 요약
- `210429`
  - 오프라인 경매 안내 페이지 레이아웃 추가 
  - 지도 추가
  
- `210430`
  - 목데이터 테스트 완료 
  - 리뷰 수정사항 반영 완료

- `210501`
  - 무한 스크롤 기능 추가 

- `210502`
  - setTimeout 추가
  - React.memo 적용 
  - styled component 이름 변경
  - Slider 공통 컴포넌트로 분리

- `210503`
  - 코드 리뷰 수정사항 반영
		 
## 수정 사항들 자세한 내용
- `210429`
  - 달력 라이브러리 사용
  - 스크롤 슬라이더 구현
  - 지도 추가 및 마커 커스텀
  - 지도 추가 시 스크롤 이벤트 추가

- `210430`
  - 목데이터 테스트 완료 
  - Theme.js 수정 완료

- `210501`
  - 무한 스크롤 기능 및 아이템 불러올때 로딩바 추가 
  
- `210502`
  - fetch 함수에 setTimeout 추가해 데이터 로드되는동안 로딩바 보여줌.
  - React.memo 적용
  - Slider.js가 메인에서도 쓰일 예정이기때문에 공통 컴포넌트로 분리.

- `210503`
  - config.js에 `BIDS_API` 분리
  - React.memo map부분만 남기고 삭제
  - 데이터 로딩 중 배경화면이 꽉차지 않는 문제 수정

## 기타 질문 및 특이 사항

## 체크 리스트 (아래 사항들이 전부 체크되어야만 merge가 됩니다!)
- [x] 필요한 test들을 완료하였고 기능이 제대로 실행되는지 확인 하였습니다.
- [x] Wecode의 코드 스타일 가이드에 맞추어 코드를 작성 하였습니다.
- [x] 제가 의도한 파일들과 수정 사항들만 커밋이 된 것을 확인 하였습니다.
- [x] 본 수정 사항들을 팀원들과 사전에 상의하였고 팀원들 모두 해당 PR에 대하여 알고 있습니다.
- [x] Git rebase와 squash를 했고 커밋 수가 하나 인것을 확인 했습니다.

Tonyk0901 commented on src/Pages/Bids/Bids.js:
```diff
@@ -0,0 +1,78 @@
+import DatePicker from 'react-horizontal-datepicker';
+import styled from 'styled-components';
+import { API, BIDS_BACKGROUND } from '../../config';
+import Theme from '../../styles/Theme';
```
- ThemeProvider 를 사용하고 있기 때문에, Theme 을 직접 import 해서 사용하지 말아 주세요!

kimnamsun replied:
```jsx
<DatePicker
          getSelectedDay={selectedDay}
          labelFormat={'MMMM'}
          endDate={90}
          color={Theme.primaryColor}
 />
```

이부분에서 사용하고 있는데 이부분도 ThemeProvider로 바꿔야할까요?

Tonyk0901 commented on src/Pages/Bids/Bids.js:
```diff
@@ -0,0 +1,78 @@
+const Header = styled.h1`
+  padding: 0.5rem;
+  font-size: ${({ type }) =>
+    type === 'title' ? '46px' : Theme.fontSizeMedium};
```
- `({ type, theme }) => theme.fontSizeMedium`

kimnamsun replied:
넵!

Tonyk0901 commented on src/Pages/Bids/Components/ScheduleList.js:
```diff
@@ -0,0 +1,150 @@
+    setIsShow(!isShow);
+  };
+
+  useEffect(() => {
+    console.log('list :', bidsList);
+  }, []);
```
- 콘솔 찍어보시려고 useEffect 를 사용하고 계신건가요?
- 함수형 컴포넌트는 결국 함수이기 때문에, 그냥 함수 내에서 console 찍어보셔도 됩니다.
```js
function func() {
 console.log('함수내에서 어디든 그냥 콘솔을 찍으셔도 됩니다.');
 return 123;
}
```

kimnamsun replied:
넵 콘솔찍는용도로 사용했는데 앞으로는 함수 내에서 찍도록 하겠습니다~!

Tonyk0901 commented on src/Pages/Bids/Components/ScheduleList.js:
```diff
@@ -0,0 +1,150 @@
+
+  useEffect(() => {
+    if (isShow) {
+      window.scrollTo(0, mapWrap.current.scrollHeight * (id * 1.5));
```
👍

Tonyk0901 commented on src/Pages/Bids/Components/Slider.js:
```diff
@@ -0,0 +1,88 @@
+  const slider = useRef();
+  const prevBtn = useRef();
+
+  useEffect(() => {
+    // console.log('slider안', arts);
+  }, []);
```
- 불필요한 useEffect 라면 삭제해주세요!

kimnamsun replied:
넵!

Tonyk0901 commented on src/Pages/Bids/Components/Slider.js:
```diff
@@ -0,0 +1,88 @@
+  };
+
+  return (
+    <>
```
- 불필요한 Fragment 입니다!

kimnamsun replied:
넵!

Tonyk0901 commented on src/styles/Theme.js:
```diff
@@ -5,10 +5,12 @@ export const flexCenter = css`
   justify-content: center;
   align-items: center;
 `;
+
 export const flexJustCenter = css`
   display: flex;
   justify-content: center;
 `;
+
 export const flexAlignCenter = css`
   display: flex;
   align-items: center;
```
- 패턴 하나 알려드립니다! 참고해보세요!
```js
const flexSet = (justify: 'center', align: 'center') => css`
  display: flex;
  justify-content: justify;
  align-items: align;
`

const theme = {
  flexSet,
};
```

Tonyk0901 reviewed (commented):
고생하셨습니다 남선님-! 리뷰 확인해주세요! 👍

kimnamsun committed:
Modify : 무한스크롤기능 구현

Modify : 코드 리팩토링 및 styled component 이름 변경

Modify : slider 공통 컴포넌트로 분리

Modify : 리뷰 수정사항 반영

Tonyk0901 commented on src/Components/Loading/Loading.js:
```diff
@@ -18,13 +18,13 @@ const loadingSpin = keyframes`
 const LoadingWrap = styled.div`
-  ${({ theme }) => theme.flexCenter};
-  min-height: 100vh;
+  ${({ theme }) => theme.flexSet('center', 'center')}
```
👍 👍

Tonyk0901 commented on src/config.js:
```diff
@@ -1,7 +1,13 @@
-export const API = 'http://10.58.6.61:8000';
+export const API = 'http://192.168.0.74:8000';
```
- 어떤 API 인지 명확하게 명시해주세요! ex) `SIGN_UP_API`
- IP 주소부는 따로 분리해주시는 게 좋습니다! ex)
```js
const IP = '192.168.0.74:8000';
const SIGN_UP_API = `http://${IP}/signup`;
```

kimnamsun replied:
api 주소만 config.js에서 관리하는 이유는 현재로썬 백엔드 아이피가 계속 바뀌기 때문에 엔드포인트마다 관리해준다면 하나하나 다 아이피를 바꿔줘야하기때문입니다,,! 그래도 다 따로 분리해야할까요?

Tonyk0901 replied:
제가 의도한 바는 지금 만약 아래와 같이 작성되어있다면, 
```js
export const API1 = 'http://192.168.0.74:8000/endpoint1';
export const API2 = 'http://192.168.0.74:8000/endpoint2';
export const API3 = 'http://192.168.0.74:8000/endpoint3';
export const API4 = 'http://192.168.0.74:8000/endpoint4';
```
백엔드 ip 가 변경될 시, API 하나하나를 일일이 수정해주어야 합니다!

IP 를 변수로 관리해주시면, 
```js
const IP = '192.168.0.74:8000';
export const API1 = 'http://${IP}/endpoint1';
export const API1 = 'http://${IP}/endpoint1';
export const API1 = 'http://${IP}/endpoint1';

```

Tonyk0901 replied:
따로 분리한다는 말에 살짝 혼동이 오신 것 같네유. 제가 말한 것은, 

지금 만약 아래와 같이 작성되어있다면, 
```js
export const API1 = 'http://192.168.0.74:8000/endpoint1';
export const API2 = 'http://192.168.0.74:8000/endpoint2';
export const API3 = 'http://192.168.0.74:8000/endpoint3';
export const API4 = 'http://192.168.0.74:8000/endpoint4';
```
백엔드 ip 가 변경될 시, API 하나하나에서 IP 주소부를 일일이 수정해주어야 합니다!

만약 IP 주소부를 변수로 관리해주시면, 
```js
const IP = '192.168.0.74:8000';
export const API1 = 'http://${IP}/endpoint1';
export const API2 = 'http://${IP}/endpoint2';
export const API3 = 'http://${IP}/endpoint3';
```
IP 가 변경될 시, IP 변수에 할당되는 값만 수정해주시면 됩니다!

Tonyk0901 commented on src/Components/Slider/Slider.js:
```diff
@@ -0,0 +1,90 @@
+          <i className="xi-angle-left-min" />
+        </DirectionBtn>
+      )}
+      {lists.map((list, index) => {
+        return type === 'bids' ? <Product key={index} art={list} /> : null;
+      })}
```
슬라이더 컴포넌트에서 Product 컴포넌트를 import 하여 사용하는 것은 재사용성의 의미에서 별로 좋지 않은 구조 같습니다. 슬라이더를 사용할 수 있는 경우가 Product 컴포넌트로 이루어진 배열로 제한되기 때문입니다. 이 부분을 children 으로 처리해주면 슬라이더를 더 많은 경우에 사용할 수 있을 것 같습니다.
```suggestion
{children}
```
이렇게 되면, Slider 를 다음과 같이 사용할 것 입니다.
```js
<Slider>
   {lists.map((list, index) => {
      return <Product key={index} art={list} />
   })}
</Slider>
```
이렇게 되면, 어떤 배열에 대해서도 슬라이더를 사용할 수 있게 됩니다.

참고해보세요!

kimnamsun replied:
넵 이렇게 하는게 더 맞겠네요 감사합니다 😊

Tonyk0901 reviewed (commented):
- 해당 Theme 부분은 import 해서 처리하는게 맞겠네요! 국 👍 
- height 가 경우에 따라서 100vh 혹은 auto 로 되도록 해주시면 될 듯 합니다. 스타일 컴포넌트 props 를 이용해 볼수 있겠네요! 
- React.memo 의 경우, 실제로 그 효과를 보기 위해서는 수백~ 수천개의 컴포넌트가 리렌더링이 되는 상황이 되어야 할겁니다. 지금과 같이 10개 이하의 컴포넌트 개수에서는 최적화 하는 것의 실질적인 효과는 없습니다.
- 그래도 공부해보자 하는 느낌으로 사용하신다면,
   1. Bids 컴포넌트에 걸어주신 React.memo 는 의미가 없습니다. Bids 컴포넌트는 bids 경로가 주어졌을 때 최상위 컴포넌트 이므로, 프롭스를 아무것도 받지 않습니다. 
   2. Map 컴포넌트의 경우 의미가 있겠네요! Map 컴포넌트에 콘솔을 작성하고, 맵 컴포넌트를 마운트 시켜둔 뒤, 상위 컴포넌트인 Bids 에서 스테이트를 변경할 시, Map 컴포넌트에 찍어둔 콘솔이 찍히지 않는 것을 확인할 수 있습니다. Map 컴포넌트로 전달되는 x, y props 값이 변하지 않았기 때문에, Map 컴포넌트는 Memoized 되어 새로 렌더링이 되지 않는 것이죠.
   3. 위와 같이 어떤 컴포넌트에서 의미가 있는지, 어떤 컴포넌트에서는 의미가 없는지 판단하시면 됩니다. React.memo 는 프롭스를 비교합니다. 만약 React.memo 로 감싸준 컴포넌트에 넘겨지는 프롭스가 자주 변화한다면, React.memo 를 사용하는 것이 오히려 리소스 낭비입니다. 프롭스가 변했는지 변하지 않았는지 비교하는 불필요한 오퍼레이션이 추가되기 때문입니다.
   4. React.memo 를 사용하시면서 주의할 점 한 가지는 프롭스로 함수가 넘겨진다면, useCallback 훅을 사용하여 함수의 참조 값을 처리해주어야 합니다. 함수형 컴포넌트에서는 컴포넌트가 렌더링 될 때마다 **함수를 항상 새로 만듭니다.** 즉, 프롭스로 넘겨지는 함수는 항상 새로운 함수 (함수의 내용에 변화가 있던지 아니던지에 상관 없이) 입니다. 그렇기 때문에 별도의 처리를 해주지 않는다면, 프롭스로 새로운 함수가 넘겨지고, React.memo 는 이를 변화로 받아드립니다. 이렇게 되면 React.memo 를 사용하는 의미가 없어집니다. 항상 재렌더링이 될테니까요. useCallback 훅은 함수의 내용이 변화가 없다면, 함수를 새로 생성하지 않고 않고 **기존에 생성된 함수를 사용하도록 해줍니다.** 이를 통해, props 로 넘겨지는 함수의 참조 값이 유지되어 React.memo 가 프롭스 변화를 제대로 인지할 수 있게됩니다.

Tonyk0901 commented:
새로 state 를 만드는 것도 방법이겠지만, state 를 만드시면 setState 하는 과정이 생길 것이고, 불필요한 렌더링이 발생될 가능성이 있습니다.

현재 존재하는 state 값만으로도 충분합니다. 예를 들어, 처음에는 데이터가 없을 것이기 때문에, data.length === 0 는 true 일 것이고, 이후에 데이터를 받아오면, data.length === 0 false 이겠네요! 

이 불리언 값을 이용하여, 조건에 따라 height 가 다르게 설정되게 하면 되지 않을까요?

kimnamsun commented:
> 새로 state 를 만드는 것도 방법이겠지만, state 를 만드시면 setState 하는 과정이 생길 것이고, 불필요한 렌더링이 발생될 가능성이 있습니다.
> 
> 현재 존재하는 state 값만으로도 충분합니다. 예를 들어, 처음에는 데이터가 없을 것이기 때문에, data.length === 0 는 true 일 것이고, 이후에 데이터를 받아오면, data.length === 0 false 이겠네요!
> 
> 이 불리언 값을 이용하여, 조건에 따라 height 가 다르게 설정되게 하면 되지 않을까요?

```jsx
<Wrapper isFirstLoading={!bidsLists.length}>
```

성훈님 감사합니당 이렇게 해서 해결했어요!!!!!!
이런식으로 하라고 하신거 맞죠?🤩
